src/sonic-ctrmgrd/ctrmgr/kube_commands.py

```diff
@@ -0,0 +1,394 @@
+    return lock_fd
+
+
+def _download_file(server, port, insecure):
+    """ Download file from Kube master to assist join as node. """
+    SERVER_ADMIN_URL = "https://{}/admin.conf"
+
+    if insecure:
+        r = urllib.request.urlopen(SERVER_ADMIN_URL.format(server),
+                context=ssl._create_unverified_context())
+    else:
+        r = urllib.request.urlopen(SERVER_ADMIN_URL.format(server))
+
+    (h, fname) = tempfile.mkstemp(suffix="_kube_join")
+    data = r.read()
+    os.write(h, data)
+    os.close(h)
+
+    # Ensure the admin.conf has given VIP as server-IP.
+    update_file = "{}.upd".format(fname)
+    cmd = 'sed "s/server:.*:{}/server: https:\/\/{}:{}/" {} > {}'.format(
+            str(port), server, str(port), fname, update_file)
+    (ret, _, err) = _run_command(cmd)
+
+    if ret != 0:
+        log_err("sed update of downloaded file failed with ret={}\n{}".
+                format(ret, err))
+        return ret
+
+    shutil.copyfile(update_file, KUBE_ADMIN_CONF)
+
+    _run_command("rm -f {} {}".format(fname, update_file))
+    log_debug("{} downloaded".format(KUBE_ADMIN_CONF))
+
+
+def _troubleshoot_tips():
+    """ log troubleshoot tips which could be handy,
+        when in trouble with join
+    """
+    msg = """
+if join fails, check the following
+
+a.  Ensure both master & node run same or compatible k8s versions
+
+b.  Check if this node already exists in master
+    Use 'sudo kubectl --kubeconfig=${KUBE_ADMIN_CONF} get nodes' to list nodes at master.
+
+    If yes, delete it, as the node is attempting a new join.
+    'kubectl --kubeconfig=${KUBE_ADMIN_CONF} drain <node name> --ignore-daemonsets'
+    'kubectl --kubeconfig=${KUBE_ADMIN_CONF} delete node <node name>'
+
+c.  In Master check if all system pods are running good.
+    'kubectl get pods --namespace kube-system'
+
+    If any not running properly, say READY column has 0/1, decribe pod for more detail.
+    'kubectl --namespace kube-system describe pod <pod name>'
+
+    For additional details, look into pod's logs.
+    @ node: /var/log/pods/<podname>/...
+    @ master: 'kubectl logs -n kube-system <pod name>'
+    """
+
+    (h, fname) = tempfile.mkstemp(suffix="kube_hints_")
+    os.write(h, str.encode(msg))
+    os.close(h)
+
+    log_error("Refer file {} for troubleshooting tips".format(fname))
+
+
+def _do_reset(pending_join = False):
+    # Drain & delete self from cluster. If not, the next join would fail
+    #
+    if os.path.exists(KUBE_ADMIN_CONF):
+        _run_command(
+                "kubectl --kubeconfig {} --request-timeout 20s drain {} --ignore-daemonsets".
+                format(KUBE_ADMIN_CONF, device_info.get_hostname()))
+
+        _run_command("kubectl --kubeconfig {} --request-timeout 20s delete node {}".
+                format(KUBE_ADMIN_CONF, device_info.get_hostname()))
+
+    _run_command("kubeadm reset -f", 10)
+    _run_command("rm -rf /etc/cni/net.d")
```

Comment:
Should not remove /etc/cni/net.d, we need 10-flannel.conflist in this dir to persist so that it can be used to simulate CNI when DUT joins and resets from master

Reply:
We will look into this in our next revision